kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,528 @@
+        }
+    }
+}
+
+/// Get vector of foreign field multiplication circuit gate types
+pub fn circuit_gates() -> Vec<GateType> {
+    vec![
+        GateType::ForeignMul0,
+        GateType::ForeignMul1,
+        GateType::ForeignMul2,
+    ]
+}
+
+/// Number of constraints for a given foreign field multiplication circuit gate type
+pub fn circuit_gate_constraint_count<F: FftField>(typ: GateType) -> u32 {
+    match typ {
+        GateType::ForeignMul0 => ForeignMul0::<F>::CONSTRAINTS,
+        GateType::ForeignMul1 => ForeignMul1::<F>::CONSTRAINTS,
+        GateType::ForeignMul2 => ForeignMul2::<F>::CONSTRAINTS,
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get combined constraints for a given foreign field multiplication circuit gate type
+pub fn circuit_gate_constraints<F: FftField>(typ: GateType, alphas: &Alphas<F>) -> E<F> {
+    match typ {
+        GateType::ForeignMul0 => ForeignMul0::combined_constraints(alphas),
+        GateType::ForeignMul1 => ForeignMul1::combined_constraints(alphas),
+        GateType::ForeignMul2 => ForeignMul2::combined_constraints(alphas),
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get the combined constraints for all foreign field multiplication circuit gate types
+pub fn combined_constraints<F: FftField>(alphas: &Alphas<F>) -> E<F> {
+    ForeignMul0::combined_constraints(alphas)
+        + ForeignMul1::combined_constraints(alphas)
+        + ForeignMul2::combined_constraints(alphas)
+}
+
+/// Foreign field multiplication CircuitGate selector polynomial
+#[serde_as]
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct SelectorPolynomial<F: FftField> {
+    /// Coefficient form
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub coeff: DP<F>,
+    /// Evaluation form (evaluated over domain d8)
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub eval8: EV<F, D<F>>,
+}
+
+/// Create foreign field multiplication circuit gates selector polynomials
+pub fn selector_polynomials<F: FftField>(
+    gates: &[CircuitGate<F>],
+    domain: &EvaluationDomains<F>,
+) -> Vec<SelectorPolynomial<F>> {
+    Vec::from_iter(circuit_gates().iter().map(|gate_type| {
+        // Coefficient form
+        let coeff = EV::<F, D<F>>::from_vec_and_domain(
+            gates
+                .iter()
+                .map(|gate| {
+                    if gate.typ == *gate_type {
+                        F::one()
+                    } else {
+                        F::zero()
+                    }
+                })
+                .collect(),
+            domain.d1,
+        )
+        .interpolate();
+
+        // Evaluation form (evaluated over d8)
+        let eval8 = coeff.evaluate_over_domain_by_ref(domain.d8);
+
+        SelectorPolynomial { coeff, eval8 }
+    }))
+}
+
+/// Switch off foreign field multiplication, with zero circuit gate selector evaluations
+pub fn off<F: Zero>() -> Vec<F> {
+    circuit_gates().iter().map(|_| F::zero()).collect()
+}
```

Comment:
I find this function a bit odd, it's short and only used in one place I believe, not sure if really necessary

Reply:
It's used in more than one place.. basically any place someone creates `ProofEvaluations` and wants foreign field multiplication switched off.  Normally you'd just have to pass a single `F::zero()` but with complex gates comprised of many `CircuitGates` like Chacha, FFMUL and Cairo you'd need to pass an array of `F::zero()` of some length, which is annoying and entails knowing internal stuff like how many `CircuitGate` there are... Thus, a helper like this makes it easy to do the right things without having to know the specific details.